src/packaging/metadata.py

```diff
@@ -406,3 +427,193 @@ def parse_email(data: Union[bytes, str]) -> Tuple[RawMetadata, Dict[str, List[st
     # way this function is implemented, our `TypedDict` can only have valid key
     # names.
     return cast(RawMetadata, raw), unparsed
+
+
+_NOT_FOUND = object()
+
+# "2.0" is technically invalid, but people used it while waiting for "2.1".
+_MetadataVersion = Literal["1.0", "1.1", "1.2", "2.0", "2.1", "2.2", "2.3"]
+
+
+class _Validator(Generic[T]):
+    name: str
+    raw_name: str
+    added: _MetadataVersion
+
+    def __init__(
+        self,
+        *,
+        added: _MetadataVersion = "1.0",
+    ) -> None:
+        self.added = added
+
+    def __set_name__(self, _owner: "Metadata", name: str) -> None:
+        self.name = name
+        self.raw_name = _RAW_TO_EMAIL_MAPPING[name]
+
+    def __get__(self, instance: "Metadata", _owner: type["Metadata"]) -> T:
+        # With Python 3.8, the caching can be replaced with functools.cached_property().
+        cache = instance.__dict__
+        value = cache.get(self.name, _NOT_FOUND)
+        if value is _NOT_FOUND:
+            try:
+                value = instance._raw[self.name]  # type: ignore[literal-required]
+            except KeyError:
+                if self.name in _STRING_FIELDS:
+                    value = ""
+                elif self.name in _LIST_FIELDS:
+                    value = []
+                elif self.name in _DICT_FIELDS:
+                    value = {}
+                else:
+                    raise InvalidMetadata(
+                        self.raw_name, f"unrecognized field: {self.raw_name!r}"
+                    )
+            try:
+                converter: Callable[[Any], T] = getattr(self, f"_process_{self.name}")
+            except AttributeError:
+                pass
+            else:
+                value = converter(value)
+            cache[self.name] = value
+            try:
+                del instance._raw[self.name]  # type: ignore[misc]
+            except KeyError:
+                pass
+        return cast(T, value)
+
+    def _invalid_metadata(self, msg: str) -> InvalidMetadata:
+        return InvalidMetadata(
+            self.raw_name, msg.format_map({"field": repr(self.raw_name)})
+        )
+
+    def _process_name(self, value: str) -> utils.NormalizedName:
+        if not value:
+            raise self._invalid_metadata("{field} is a required field")
+        return utils.canonicalize_name(value, validate=True)
+
+    def _process_version(self, value: str) -> version_module.Version:
+        if not value:
+            raise self._invalid_metadata("{field} is a required field")
+        return version_module.parse(value)
+
+    def _process_summary(self, value: str) -> str:
+        """Check the field contains no newlines."""
+        if "\n" in value:
+            raise self._invalid_metadata("{field} must be a single line")
+        return value
+
+    def _process_metadata_version(self, value: str) -> _MetadataVersion:
+        if value not in {"1.0", "1.1", "1.2", "2.0", "2.1", "2.2", "2.3"}:
```

Comment:
nit: Consider pulling this out into a global, around the `_MetadataVersion` global.